java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+ * This state may be achieved two ways:
+ *  1. If the source of the taint is a `File.createTempFile(_, _)` call.
+ *  2. From state `FromSystemPropertyFlowState` if `File.createTempFile(_, _, taint)` is called.
+ */
+private class FromFileCreateTempFileFlowState extends DataFlow::FlowState {
+  FromFileCreateTempFileFlowState() { this = "FromFileCreateTempFileFlowState" }
+}
+
+/**
+ * The flow state after the `File::delete` method has been called on tracked taint.
+ * This is the final state of the flow before hijackable call `File::mkdir(s)` (ie. the sink) is reached.
+ */
+private class FromDeleteFileFlowState extends DataFlow::FlowState {
+  FromDeleteFileFlowState() { this = "FromDeleteFileFlowState" }
+}
+
+/**
+ * The flow state after the `File::mkdir(s)` method has been called on tracked taint.
+ * This is the final state, and is only used when computing the 'unsafe' call site.
+ */
+private class FromMkdirsFileFlowState extends DataFlow::FlowState {
+  FromMkdirsFileFlowState() { this = "FromMkdirsFileFlowState" }
+}
+
+/**
+ * A guard that is checking if a directory exists, throwing if it does not exist.
+ */
+private Guard directoryExistsGuard(ThrowStmt s) {
+  result =
+    any(MethodAccess existanceCheck |
+      (
+        existanceCheck.getMethod() instanceof MethodFileExists
+        or
+        existanceCheck.getMethod() instanceof MethodFileIsDirectory
+      )
+    ) and
+    result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is verifying that the directory is sucsessfully created, throwing when it is not created.
+ */
+private Guard directoryCreationGuard(ThrowStmt s) {
+  result = any(MethodAccess creationCheck |
+    creationCheck.getMethod() instanceof MethodFileCreatesDirs
+  ) and
+  result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is safely verfying that a directory is created.
+ * 
+ * 'Safe' means that the directory is guarnteed to have been created, and the directory did not already exist.
+ */
+private Guard safeDirectoryCreationGuard(ThrowStmt s) {
+  result = directoryCreationGuard(s) and
+  // This guard is not 'safe' if a `directoryExistsGuard` is also guarding this throw statement
+  not exists(directoryExistsGuard(s))
+}
+
+/**
+ * An expression that will create a directory without throwing an exception if a file/directory already exists.
+ */
+private predicate isNonThrowingDirectoryCreationExpression(
+  Expr fileInstanceExpr, MethodAccess creationCall
+) {
+  creationCall.getMethod() instanceof MethodFileCreatesDirs and
+  creationCall.getQualifier() = fileInstanceExpr and
+  // Check for 'throwing creations' and ensure that this call is not used in a throwing case.
+  if creationCall.(Guard).directlyControls(_, _) then // The creation call is used in a guard
+    not creationCall = safeDirectoryCreationGuard(_) // Ensure that the guard is insufficient.
+  else any() // The creation call is not used in a guard. Thus is not a 'throwing creation'.
```

Comment:
Wondering if there's a better way to write this. Suggestions welcome.

Reply:
`creationCall.(Guard).directlyControls(_, _) implies not creationCall = safeDirectoryCreationGuard(_)`?

Reply:
You know, I was trying to figure out how to use `implies` but I still have difficult wrapping my head around it. I'll give this a shot, thanks.